src/main/java/io/gravitee/policy/jwt/jwks/URLJWKSourceResolver.java

```diff
@@ -51,23 +54,36 @@ public URLJWKSourceResolver(TemplateEngine templateEngine, String url, ResourceR
 
     @Override
     public CompletableFuture<JWKSource<C>> resolve() {
-        JWKSource<C> jwkSource = cache.getIfPresent(jwksUrl.toString());
 
-        if (jwkSource == null) {
-            return resourceRetriever
-                    .retrieve(jwksUrl)
-                    .thenCompose(resource -> {
-                        try {
-                            JWKSet parsedJWKSet = JWKSet.parse(resource.getContent());
-                            ImmutableJWKSet<C> immutableJWKSet = new ImmutableJWKSet<>(parsedJWKSet);
-                            cache.put(jwksUrl.toString(), immutableJWKSet);
-                            return CompletableFuture.completedFuture(immutableJWKSet);
-                        } catch (ParseException e) {
-                            return CompletableFuture.completedFuture(null);
-                        }
-                    });
+        CachedJWKSource cachedJWKSource = cache.get(jwksUrl.toString());
+        if (cachedJWKSource != null && !isCacheExpired(cachedJWKSource)) {
```

Comment:
maybe the new `CachedJWKSource` can have a method `isExpired()`
```suggestion
        if (cachedJWKSource != null && !cachedJWKSource.isExpired()) {
```

Reply:
As discussed together, it's not the aim of a cached object to know is retention period.
Maybe we could extends the Map, with something like RefereshingConcurrentHashMap<T>, to handle the expiration delay in it. But it's actually overdesign.